### DISCUSSION TITLE
Multithreading options

### DISCUSSION THREAD

**Discussion by ateixeira0163** [sofa-framework/sofa#2913] [Write / Run a simulation] [answered]
Hi there!
I'm currently using the plugin `Multithreading` from SOFA and I made some changes in my script to put these objects in my scene:
Added `AnimationLoopParallelScheduler` to the `root` with `threadNumber=0`. (I kept `FreeMotionAnimationLoop` for the friction contact constraint).
Changed the `multithreading=True` in the `GenericConstraintSolver`
Changed `BruteForceBroadPhase` to `ParallelBruteForceBroadPhase`
Changed `BVHNarrowPhase` to `ParallelBVHNarrowPhase`
Changed `HexahedronFEMForceField` to `ParallelHexahedronFEMForceField`
And still, after those changes, an example simulation that took 19h22m with a single thread took 19h12 with multithreading. Is it normal that the difference is so low (I made sure that there was nothing else in the computer and the computer has 8 threads total). 

Thanks for all information possible :) 
Ps.: @hugtalbot asked me to tag you @alxbilger

**Comment by alxbilger** [comment, marked as answer]
Hi,

if I understand correctly, your goal is not to use the MultiThreading plugin, but to speed up your simulation. In that case, I suggest you to read this doc page: https://github.com/sofa-framework/doc/blob/master/15_Using_SOFA/32_Performances/20_Improve_performances.md. The most important thing is to identify the bottleneck. Otherwise, we won't be able to help. For example, you use `ParallelBruteForceBroadPhase`, but if you have collision only between two objects, it is counter-productive.

The multithreading plugin is not very used, and not a lot of effort has been made to optimize it. Consider it experimental, without any guarantee of improving the performances. Still, I encourage you to read the documentation: https://github.com/sofa-framework/doc/blob/master/35_Plugins/50_Usual_plugins/80_MultiThreading.md. In particular, using `ParallelHexahedronFEMForceField` speeds up the simulation only in specific cases.

Again, identifying the bottleneck is the key.

Alex

**Reply by ateixeira0163** [reply]
I only have one object with self-contact. 
Thanks! I'm gonna read to optimize my simulation!
